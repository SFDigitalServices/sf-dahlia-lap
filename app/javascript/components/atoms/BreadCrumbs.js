import React from 'react'
import PropTypes from 'prop-types'

const Item = ({ item, current }) => {
  if (current) {
    return (
      <li className='current'>
        <a href={item.link} aria-current='page'>
          {item.title}
        </a>
      </li>
    )
  } else {
    return (
      <li>
        <a href={item.link}>{item.title}</a>
      </li>
    )
  }
}

const BreadCrumbs = ({ items }) => {
  if (!items || items.length === 0) return null

  const lastItem = items[items.length - 1]
  return (
    <nav aria-label='breadcrumb'>
      <ol className='breadcrumbs'>
        {items.slice(0, items.length - 1).map((item) => (
          <Item key={item.title} item={item} />
        ))}
        <Item key={lastItem.name} item={lastItem} current />
      </ol>
    </nav>
  )
}

Item.propTypes = {
  item: PropTypes.shape({
    title: PropTypes.string.isRequired,
    link: PropTypes.string.isRequired
  })
}

BreadCrumbs.propTypes = {
  items: PropTypes.array.isRequired
}

export default BreadCrumbs
