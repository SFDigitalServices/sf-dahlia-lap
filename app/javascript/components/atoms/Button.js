import React from 'react'

import classNames from 'classnames'
import PropTypes from 'prop-types'

import { toPx } from 'utils/cssUtils'

const buttonSizeStyles = (clearVerticalPadding, minWidth) => {
  const widthStyle = minWidth
    ? {
        minWidth: toPx(minWidth)
      }
    : {}

  const heightStyle = clearVerticalPadding
    ? {
        paddingTop: 0,
        paddingBottom: 0
      }
    : {}

  return {
    ...heightStyle,
    ...widthStyle
  }
}

const buttonChildrenStyles = () => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  width: '100%'
})

const textStyles = (hasLeftIcon, hasRightIcon) => ({
  marginTop: '1rem',
  marginBottom: '1rem',
  marginRight: hasRightIcon && '1rem',
  marginLeft: hasLeftIcon && '1rem',
  flexGrow: 1
})

const iconWrapperStyles = () => ({
  flexShrink: 0,

  // we need to override font-size because of a css button selector that sets
  // the font size for all children. This is a problem because it adds 3px
  // to any icon wrapper.
  fontSize: 0
})

const wrapWithStyle = (node, style) => node && <span style={style}>{node}</span>

const Button = ({
  classes = null,
  iconLeft = null,
  iconRight = null,
  noBottomMargin,
  paddingHorizontal = 'normal',
  small = false,
  tertiary = false,
  text = null,
  tiny = false,
  type = 'button',
  minWidthPx = null,
  style = {},
  ...rest
}) => {
  const btnClassNames = classNames(classes, {
    'tight-padding': paddingHorizontal === 'tight',
    'extra-padding': paddingHorizontal === 'extra',
    small: small,
    tertiary: tertiary,
    tiny: tiny,
    'margin-bottom-none': noBottomMargin
  })

  const hasText = !!text

  return (
    <button
      className={btnClassNames}
      type={type}
      style={{ ...buttonSizeStyles(hasText, minWidthPx), ...style }}
      {...rest}
    >
      <div style={buttonChildrenStyles()}>
        {wrapWithStyle(iconLeft, iconWrapperStyles())}
        {wrapWithStyle(text, textStyles(!!iconLeft, !!iconRight))}
        {wrapWithStyle(iconRight, iconWrapperStyles())}
      </div>
    </button>
  )
}

Button.propTypes = {
  type: PropTypes.oneOf(['button', 'submit']),
  noBottomMargin: PropTypes.bool,
  classes: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
  iconLeft: PropTypes.node,
  iconRight: PropTypes.node,
  minWidthPx: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  paddingHorizontal: PropTypes.oneOf(['normal', 'tight', 'extra']),
  small: PropTypes.bool,
  style: PropTypes.object,
  tertiary: PropTypes.bool,
  text: PropTypes.node,
  tiny: PropTypes.bool
}

export default Button
