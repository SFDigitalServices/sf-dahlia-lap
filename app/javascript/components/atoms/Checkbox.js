import React from 'react'

import classNames from 'classnames'
import { PropTypes } from 'prop-types'

const Checkbox = ({ id, checked = false, indeterminate = false, onClick = () => {} }) => (
  <>
    <input
      id={id}
      type='checkbox'
      checked={indeterminate || checked}
      onClick={onClick}
      // onChange handler required because we're setting 'checked' prop,
      // but we actually want our handler to be onClick instead,
      // so this is a no-op function.
      onChange={() => {}}
      className={classNames('no-margin', { indeterminate: indeterminate })}
    />
    <label style={{ margin: '0px' }} className='form-label' htmlFor={id} />
  </>
)

Checkbox.propTypes = {
  id: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  indeterminate: PropTypes.bool,
  onClick: PropTypes.func
}
export default Checkbox
