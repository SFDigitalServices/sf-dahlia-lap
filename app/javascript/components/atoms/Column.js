import React from 'react'

const Column = ({ children, span, end, form }) => (
  <div
    className={`columns ${span ? `small-${span}` : ''} ${end ? 'end' : ''} ${
      form ? 'form-grid_item' : ''
    }`}
  >
    {children}
  </div>
)

export default Column
