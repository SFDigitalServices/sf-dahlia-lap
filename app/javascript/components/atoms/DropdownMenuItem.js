import React from 'react'

const DropdownMenuItem = ({ value, label, selected, onChange, onKeyDown }) => {
  let liClassName = 'dropdown-menu_item'
  if (selected)
    liClassName += ' is-selected'

  return (
    <li className={liClassName} role="option" aria-selected={selected} >
      <a tabIndex="0" onKeyDown={(e) => onKeyDown(e, value, label)} onClick={(e) => onChange(e, value, label) }>{label}</a>
    </li>
  )
}

export default DropdownMenuItem
