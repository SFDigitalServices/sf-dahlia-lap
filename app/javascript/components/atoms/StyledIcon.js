import React from 'react'
import PropTypes from 'prop-types'

import { toRem } from '~/utils/cssUtils'

const DEFAULT_ICON_SIZE_REM = '1.5rem'

const getSizeStyle = (sizeRem) => ({
  width: sizeRem,
  height: sizeRem
})

const ICON_STYLE_REGULAR = getSizeStyle(DEFAULT_ICON_SIZE_REM)
const ICON_STYLE_MEDIUM = getSizeStyle('1rem')
const ICON_STYLE_SMALL = getSizeStyle('.75rem')
const ICON_STYLE_TINY = getSizeStyle('.5rem')

const getCustomSizeStyle = (customSize) => {
  let customSizeRem = toRem(customSize)

  if (customSizeRem === null) {
    console.error(`Invalid size value: ${customSize}`)
    customSizeRem = DEFAULT_ICON_SIZE_REM
  }

  return getSizeStyle(customSizeRem)
}

const getIconStyle = (size, customSize) => {
  switch (size) {
    case 'default':
      return ICON_STYLE_REGULAR
    case 'small':
      return ICON_STYLE_SMALL
    case 'medium':
      return ICON_STYLE_MEDIUM
    case 'tiny':
      return ICON_STYLE_TINY
  }
}

const StyledIcon = ({ size = 'default', customSizeRem = null, icon }) => (
  <svg style={customSizeRem ? getCustomSizeStyle(customSizeRem) : getIconStyle(size)}>
    <use xlinkHref={`#i-${icon}`} />
  </svg>
)

StyledIcon.propTypes = {
  icon: PropTypes.string.isRequired,
  size: PropTypes.oneOf(['default', 'small', 'medium', 'tiny']),
  customSizeRem: PropTypes.string
}

export default StyledIcon
