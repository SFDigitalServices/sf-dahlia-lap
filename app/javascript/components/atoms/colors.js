// These colors are pulled from the sf-dahlia-pattern-library
// https://github.com/SFDigitalServices/sf-dahlia-pattern-library/blob/main/components/01-atoms/color.config.yml
// Whenever possible, we should use classnames to style colors but we added these in case we need them.
export const COLORS = {
  'dust': '#f5f8f9',
  'snow': '#f9f9f9',
  'mist': '#f7f7f7',
  'vapor': '#f6f6f6',
  'wash': '#efefef',
  'smoke': '#dedee0',
  'iron': '#ccc',
  'base': '#aaa',
  'aluminum': '#999',
  'steel': '#767676',
  'charcoal': '#555',
  'tuatara': '#30383a',
  'oil': '#333',
  'ebony': '#242c2e',
  'jet': '#222',
  'primary': '#0177DA',
  'deep': '#0067be',
  'dark': '#0d4b80',
  'primaryTint': '#daeeff',
  'attention': '#00bed5',
  'attentionTint': '#c8f1ff',
  'royal': '#b85ed5',
  'royalTint': '#eed7f5',
  'splash': '#ff6627',
  'splashTint': '#ffd2c0',
  'lush': '#99cd00',
  'lushTint': '#f2ffcd',
  'alert': '#e31c3d',
  'alertTint': '#f9d2d8',
  'success': '#2e8540',
  'successTint': '#b4e5be',
  'warn': '#fdb81e',
  'warnTint': '#fee8b6'
}
