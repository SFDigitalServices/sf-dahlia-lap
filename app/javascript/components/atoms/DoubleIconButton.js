import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import Button from './Button'
import Icon from './Icon'

/**
 * button that has text and buttons on either side of it.
 */
const DoubleIconButton = ({
  type = 'button',
  text = null,
  tertiary,
  noBottomMargin,
  classes,
  leftIcon = '',
  rightIcon = '',
  ...rest
}) => {
  const btnClassNames = classNames(classes, 'has-icon--both-sides')
  return (
    <Button
      classes={btnClassNames}
      type={type}
      tertiary={tertiary}
      noBottomMargin={noBottomMargin}
      {...rest}
    >
      <div style={{ display: 'flex' }}>
        <Icon icon={leftIcon} uiIconClassName='ui-icon--left' />
        {text}
        <Icon icon={rightIcon} uiIconClassName='ui-icon--right' />
      </div>
    </Button>
  )
}

DoubleIconButton.propTypes = {
  type: PropTypes.oneOf(['button', 'submit']),
  noBottomMargin: PropTypes.bool
}

DoubleIconButton.defaultProps = {
  type: 'button',
  noBottomMargin: false
}

export default DoubleIconButton
