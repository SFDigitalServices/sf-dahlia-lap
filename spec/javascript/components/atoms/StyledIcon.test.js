import React from 'react'

import { render } from '@testing-library/react'

import { COLORS } from 'components/atoms/colors'
import StyledIcon from 'components/atoms/StyledIcon'

describe('StyledIcon', () => {
  describe('with default props', () => {
    test('should render the correct icon and apply the default style', () => {
      const { asFragment } = render(<StyledIcon icon='list-unordered' />)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('with size prop = small', () => {
    test('should apply the small style', () => {
      const { asFragment } = render(<StyledIcon size='small' icon='list-unordered' />)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('with custom size', () => {
    test('should apply the custom size style', () => {
      const { asFragment } = render(<StyledIcon customSizeRem='3rem' icon='list-unordered' />)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('with custom fill', () => {
    test('should apply the custom fill style', () => {
      const { asFragment } = render(
        <StyledIcon customFill={COLORS.success} icon='list-unordered' />
      )
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
