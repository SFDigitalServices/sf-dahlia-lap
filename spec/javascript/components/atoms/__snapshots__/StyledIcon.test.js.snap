// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyledIcon with custom fill should apply the custom fill style 1`] = `
<DocumentFragment>
  <svg
    style="width: 1.5rem; height: 1.5rem; fill: #2e8540;"
  >
    <use
      style="fill: #2e8540;"
      xlink:href="#i-list-unordered"
    />
  </svg>
</DocumentFragment>
`;

exports[`StyledIcon with custom size should apply the custom size style 1`] = `
<DocumentFragment>
  <svg
    style="width: 3rem; height: 3rem;"
  >
    <use
      xlink:href="#i-list-unordered"
    />
  </svg>
</DocumentFragment>
`;

exports[`StyledIcon with default props should render the correct icon and apply the default style 1`] = `
<DocumentFragment>
  <svg
    style="width: 1.5rem; height: 1.5rem;"
  >
    <use
      xlink:href="#i-list-unordered"
    />
  </svg>
</DocumentFragment>
`;

exports[`StyledIcon with size prop = small should apply the small style 1`] = `
<DocumentFragment>
  <svg
    style="width: .75rem; height: .75rem;"
  >
    <use
      xlink:href="#i-list-unordered"
    />
  </svg>
</DocumentFragment>
`;
